src/isolation/fd.rs

```diff
@@ -1,33 +1,107 @@
 use std::{
-	collections::HashSet,
+	fmt,
 	os::fd::{FromRawFd, OwnedFd, RawFd},
 };
 
-#[derive(Default, Debug)]
+use slab::Slab;
+use yoke::{Yoke, erased::ErasedArcCart};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct GuestFd(pub i32);
+
+impl fmt::Display for GuestFd {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "GuestFd({})", self.0)
+	}
+}
+
+impl From<usize> for GuestFd {
+	fn from(x: usize) -> Self {
+		// The following should never panic unless someone allocates more than i32::MAX file descriptors
+		Self(x.try_into().unwrap())
+	}
+}
+
+impl GuestFd {
+	fn is_standard(self) -> bool {
+		self.0 < 3
+	}
+
+	fn get(self) -> usize {
+		self.0.try_into().unwrap()
+	}
+}
+
+/// Description of what this guest file descriptor wraps:
+#[derive(Clone)]
+pub enum FdData {
+	/// A host file descriptor
+	Raw(RawFd),
+
+	#[allow(dead_code)]
+	/// An in-memory slice (possibly mmap-ed)
+	///
+	/// SAFETY: It is not allowed for `data` to point into guest memory.
+	Virtual {
+		data: Yoke<&'static [u8], ErasedArcCart>,
+		offset: u64,
+	},
+}
+
+impl fmt::Debug for FdData {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			FdData::Raw(r) => write!(f, "Raw({})", r),
+			FdData::Virtual { data, offset } => {
+				let data = data.get();
+				let data_snip = &data[..core::cmp::min(10, data.len())];
+				write!(f, "Virtual({:?} @ {})", data_snip, offset)
+			}
+		}
+	}
+}
+
+impl FromRawFd for FdData {
+	#[inline(always)]
+	unsafe fn from_raw_fd(fd: RawFd) -> Self {
+		Self::Raw(fd)
+	}
+}
+
+#[derive(Debug)]
 pub struct UhyveFileDescriptorLayer {
-	fdset: HashSet<RawFd>,
+	fds: Slab<FdData>,
 }
 
-impl UhyveFileDescriptorLayer {
-	/// Returns all present file descriptors.
-	pub(crate) fn get_fds(&self) -> Vec<&i32> {
-		self.fdset.iter().collect::<Vec<_>>()
+impl Default for UhyveFileDescriptorLayer {
+	fn default() -> Self {
+		let mut fds = Slab::new();
+		// fill the first 3 fds (0, 1, 2) = the standard streams to avoid weird effects
+		fds.insert(FdData::Raw(0));
```

Comment:
I find the approach of inserting some data and then ignoring certain "dangerous" operations using `is_default` to be a good idea.